### ISSUE TITLE
$ not support in the name of query parameters

### ISSUE THREAD

**Issue by pvreeburg** [OpenFeign/feign#1036]
I encountered an issue with the processing of query parameters that contain a $ in their name. These apparently aren't processed as expressions but as literals.

I'm currently using FeignJaxRs2 versions '10.2.3'.

Example :
```java
public interface TestClient {
    @GET
    @Path("/test")
    String apiCall(@QueryParam("$fields") String value);
}

public void test_with_empty_name() {
    String result = testClient.apiCall("field1,field2");
}
```
results in the URI :

.../test?$fields={$fields}

instead of:

.../test?$fields=field1,field2

It appears to go wrong in the class feign.template.Expressions.
The regex used to validate the name of the queryparameter doesn't currently allow $.

**Comment by kdavisk6** [comment]
TL;DR - Please see #894 for an explanation and a possible workaround for now.

Explanation:
Part of the changes we've made in the past few releases have changed the way that we recognize and support uri template expressions.  In practice this means that template expression variable names that include reserved characters, as defined in [RFC 6570](https://tools.ietf.org/html/rfc6570#section-1.5), may result in unexpected behavior.  The RFC states that those values should be [pct-encoded](https://en.wikipedia.org/wiki/Percent-encoding), but we don't enforce that.

If time permits, or others want to help, we can look into relaxing these restrictions more, but I'm afraid that there are more use cases that others are now depending on that will cause grief in other areas.  

I know it's not an *real* answer, but it's an honest one.
